quil-py/quil/program/__init__.pyi

```diff
@@ -175,6 +175,121 @@ class Program:
+
+                from quil.program import Program
+
+                program = Program.parse("CZ 0 1; CZ 0 2")
+
+                print(program.to_quil())
+
+                control_flow_graph = program.control_flow_graph()
+                assert control_flow_graph.has_dynamic_control_flow() == False
+
+                basic_blocks = control_flow_graph.basic_blocks()
+                assert len(basic_blocks) == 1
+
+                schedule = blocks[0].as_fixed_schedule(program, False)
+                print(f"Duration = {schedule.duration()}")
+
+                print(schedule.items())
+
+
+         To understand why `include_zero_duration_instructions` is useful, consider a program like this:
+
+         .. example-code::
+
+            .. code-block:: text
+
+                # Two-qubit frame
+                DEFFRAME 0 1 "a":
+                    ATTRIBUTE: 1
+
+                # One-qubit frame
+                DEFFRAME 1 "a":
+                    ATTRIBUTE: 1
+
+                DEFCAL A 0:
+                    PULSE 1 "a" flat(duration: 1.0)
+
+                # The FENCE "plays" immediately upon program start at time 0
+                # The `PULSE` is blocked by the `PULSE` of `A 0` on intersecting frame 1 "a", and does not play until time 1
+                DEFCAL B 0:
+                    FENCE 0
+                    PULSE 0 1 "a" flat(duration: 1.0)
+
+                A 0
+                B 0
+
+        If ``include_zero_duration_instructions`` is ``True``, ``B 0`` will be scheduled from time 0 to time 2, because it includes the time
+        of the FENCE.
+
+        If not, it will be from time 1 to time 2, considering only the non-zero duration instructions (here, ``PULSE``) within the calibration.
+        """
+    def gate_depth(self, gate_minimum_qubit_count: int) -> int:
+        """
+        Returns the length of the longest path from an initial instruction (one with no prerequisite instructions) to a final
+        instruction (one with no dependent instructions), where the length of a path is the number of gate instructions in the path.
+
+        :param gate_minimum_qubit_count: The minimum number of qubits in a gate for it to be counted in the depth.
+        """
+    def gate_volume(self) -> int:
+        """
+        Return the total number of gates in the block.
+        """
+    def label(self) -> Optional[Target]:
+        """
+        Return the label of the block, if any. This is used to target this block in control flow.
+        """
+    def instructions(self) -> List[Instruction]:
+        """
+        Return a list of the instructions in the block, in order of definition.
+
+        This does not include the label or terminator instructions.
+        """
+    def terminator(self) -> Optional[Instruction]:
+        """
+        Return the control flow terminator instruction of the block, if any.
+
+        If this is ``None``, the implicit behavior is to "continue" to the subsequent block.
```

Comment:
I feel like a little more explanation is required here; it took me a moment to deconflict concepts from the Wikipedia article on CFGs to what is happening in Quil. Each basic block can have multiple output arcs- in Quil it's probably limited to 2 (or 1 and it continues to the next block) and this happens only on conditional jump instructions? It would be worth saying that here, and it would be more semantically rich if there was a Quil control flow subclass of `Instruction` that was the return type here that helped you see it was only jump instructions.

Reply:
Also, is this not by definition `self.instructions[-1]`? Or is the terminator excluded from `instructions`? If included, I can see why you would still have this method, because the naming provides a sort of semantic richness that is valuable (ie. same argument for having the "end" method on the span; its sort of canonical to the definition).

Reply:
Great; a distinction worth documenting?